src/raftstore/store/peer.rs

```diff
@@ -1445,88 +1349,79 @@ impl Peer {
+                // The request could be proposed when the peer was leader.
+                // TODO: figure out that it's necessary to notify stale or not.
+                let term = self.term();
+                apply::notify_stale_req(term, cb);
+            }
+        }
+    }
+
     /// Responses to the ready read index request on the replica, the replica is not a leader.
     fn post_pending_read_index_on_replica<T, C>(&mut self, ctx: &mut PollContext<T, C>) {
-        if self.pending_reads.ready_cnt > 0 {
-            for _ in 0..self.pending_reads.ready_cnt {
-                let mut read = self.pending_reads.reads.pop_front().unwrap();
-                RAFT_READ_INDEX_PENDING_COUNT.sub(read.cmds.len() as i64);
-                assert!(read.read_index.is_some());
-
-                let is_read_index_request = read.cmds.len() == 1
-                    && read.cmds[0].0.get_requests().len() == 1
-                    && read.cmds[0].0.get_requests()[0].get_cmd_type() == CmdType::ReadIndex;
-
-                let term = self.term();
-                if is_read_index_request {
-                    debug!("handle reads with a read index";
-                        "request_id" => ?read.binary_id(),
-                        "region_id" => self.region_id,
-                        "peer_id" => self.peer.get_id(),
-                    );
-                    for (req, cb) in read.cmds.drain(..) {
-                        cb.invoke_read(self.handle_read(ctx, req, true, read.read_index));
-                    }
-                    self.pending_reads.ready_cnt -= 1;
-                } else if self.ready_to_handle_unsafe_replica_read(read.read_index.unwrap()) {
-                    debug!("handle reads with a read index";
-                        "request_id" => ?read.binary_id(),
-                        "region_id" => self.region_id,
-                        "peer_id" => self.peer.get_id(),
-                    );
-                    for (req, cb) in read.cmds.drain(..) {
-                        // We should check epoch since the range could be changed
-                        if req.get_header().get_replica_read() {
-                            cb.invoke_read(self.handle_read(ctx, req, true, read.read_index));
-                        } else {
-                            apply::notify_stale_req(term, cb);
-                        }
-                    }
-                    self.pending_reads.ready_cnt -= 1;
-                } else {
-                    self.pending_reads.reads.push_front(read);
-                }
+        while let Some(read) = self.pending_reads.pop_front() {
+            assert!(read.read_index.is_some());
+            let is_read_index_request = read.cmds.len() == 1
+                && read.cmds[0].0.get_requests().len() == 1
+                && read.cmds[0].0.get_requests()[0].get_cmd_type() == CmdType::ReadIndex;
+
+            if is_read_index_request {
+                self.response_read(read, ctx, false);
```

Comment:
Is it necessary to response read index request first if it's in the middle of the queue?

Reply:
Let's resolve it later.